pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,466 @@
+) error {
+	nodes := args.(DiskStallArgs).Nodes
+	return s.WaitForSQLLiveness(ctx, l, nodes, 3*time.Minute)
+}
+
+type throughput struct {
+	limited        bool
+	bytesPerSecond string
+}
+
+type bandwidthType int8
+
+const (
+	readBandwidth bandwidthType = iota
+	writeBandwidth
+)
+
+func (rw bandwidthType) cgroupV2BandwidthProp() string {
+	switch rw {
+	case readBandwidth:
+		return "rbps"
+	case writeBandwidth:
+		return "wbps"
+	default:
+		panic("unreachable")
+	}
+}
+
+func (s *CGroupDiskStaller) setThroughput(
+	ctx context.Context,
+	l *logger.Logger,
+	readOrWrite []bandwidthType,
+	bw throughput,
+	nodes install.Nodes,
+	cockroachIOController string,
+) error {
+	maj, min, err := s.DiskDeviceMajorMinor(ctx, l)
+	if err != nil {
+		return err
+	}
+
+	var limits []string
+	for _, rw := range readOrWrite {
+		bytesPerSecondStr := "max"
+		if bw.limited {
+			bytesPerSecondStr = bw.bytesPerSecond
+		}
+		limits = append(limits, fmt.Sprintf("%s=%s", rw.cgroupV2BandwidthProp(), bytesPerSecondStr))
+	}
+	l.Printf("setting cgroup bandwith limits:\n%v", limits)
+
+	return s.Run(ctx, l, nodes, "sudo", "/bin/bash", "-c", fmt.Sprintf(
+		`'echo %d:%d %s > %s'`,
+		maj,
+		min,
+		strings.Join(limits, " "),
+		cockroachIOController,
+	))
+}
+
+func (s *CGroupDiskStaller) GetReadWriteBytes(
+	ctx context.Context, l *logger.Logger, node install.Nodes,
+) (int, int, error) {
+	maj, min, err := s.DiskDeviceMajorMinor(ctx, l)
+	if err != nil {
+		return 0, 0, err
+	}
+	// Check the number of bytes read and written to disk.
+	res, err := s.RunWithDetails(
+		ctx, l, node,
+		fmt.Sprintf(`grep -E '%d:%d' /sys/fs/cgroup/system.slice/io.stat |`, maj, min),
+		`grep -oE 'rbytes=[0-9]+|wbytes=[0-9]+' |`,
+		`awk -F= '{printf "%s ", $2} END {print ""}'`,
+	)
+	if err != nil {
+		return 0, 0, err
+	}
+	fields := strings.Fields(res.Stdout)
+	if len(fields) != 2 {
+		return 0, 0, errors.Errorf("expected 2 fields, got %d: %s", len(fields), res.Stdout)
+	}
+
+	readBytes, err := strconv.Atoi(fields[0])
+	if err != nil {
+		return 0, 0, err
+	}
+	writeBytes, err := strconv.Atoi(fields[1])
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return readBytes, writeBytes, nil
+}
+
+type DmsetupDiskStaller struct {
+	GenericFailure
+}
+
+func MakeDmsetupDiskStaller(
+	clusterName string, l *logger.Logger, secure bool,
+) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+
+	genericFailure := GenericFailure{c: c, runTitle: DmsetupDiskStallName}
+	return &DmsetupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+func (s *DmsetupDiskStaller) Description() string {
+	return "dmsetup disk staller"
+}
+
+func (s *DmsetupDiskStaller) Setup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	diskStallArgs := args.(DiskStallArgs)
+	var err error
+
+	// Disabling journaling requires the cockroach process to not have been started yet.
+	if diskStallArgs.RestartNodes {
```

Comment:
Note to self: having to restart the cluster to setup/cleanup dmsetup seems really disruptive - makes me lean towards using cgroups in the failure injection framework